### PR TITLE
Alexander/config

### LINKED ISSUE #110: Configuration: Create configuration class
Referencing #103 for design.

This file, `configuration.py' will be a Python script that holds the class `Configuration` for holding all the data taken in by the XML file in #103. 

The data from the XML file should be loaded into a dictionary. Not only does this allow for quicker access of information but it also separates the manipulation of configuration data from the configuration file.  The loading of the XML file is covered in #111

### LINKED ISSUE #111: Configuration: Be able to load XML files into dictionary
Referencing #103 for design.

When given an XML file, we should be able to parse it into a dictionary for easy access by the program. 
Should also be able to check the XML file to make sure all required information is there.

# XML Parser
## ElementTree
Odds are, [ElementTree](https://docs.python.org/2.7/library/xml.etree.elementtree.html) is the best bet, being built into python and whatnot. However, we might need to address potential [XML Vulnerabilities ](https://docs.python.org/2.7/library/xml.html#xml-vulnerabilities) that might appear.   
Another cool thing about ElementTree is it does have a C implementation `xml.etree.cElementTree` that might be helpful when we roll out the distributed computing platform.

## LXML
If we want to get fancy though, [LXML](http://lxml.de) seems to be a much better XML parser. LXML is a wrapper around some C libraries so again, when we roll out the distributed computing platform, and if we need to switch the configuration loader to a C/C++ program, things _might_ go smoother. LXML also has XLST support, which we have discussed possibly bringing into the program in the future. I do not believe that ElementTree does.

### LINKED ISSUE #121: Configuration: Saving data
See #103 for design discussion.

Parts of the configuration data will be manipulated by the program and will then need to be saved back to an XML file.

The way the program handles configuration data now is described in #110.

It is important to note that while _technically_ any part of the configuration data can be modified by the program, there are only some parts of the configuration data that we actually wanted edited. With the way the configuration class is being built, this will actually be pretty easy to handle.

`configuration.py` should contain a method that will save or output the data back to an XML file. This should be the same XML file that gets passed in. Now the way things work, the XML data that is loaded into the program is never modified by the program up to this point. Instead, all the data is loaded into a dictionary object and that is what gets modified by the program. Therefore in order to to save the data we want saved back to the XML file, the save method should only update the XML file from the keys of the dictionary object that we deem "modifiable".

Meaning, if for whatever reason the program accidentally changes the project name value, that won't get reflected in the output XML file.

Things that can be modified by the program:
- Models

### PR DESCRIPTION
**Pull request by asclines**
Welcome the new configuration tool to the family!

# Configuration
## Overview
This pull request brings in the solution for getting user input as discussed on and designed by the team in #103 
With this way of handling configuration, the user will be able to 'feed' in an XML file into the pipeline that will hold _all_ of their input for the program. This serves as a one-stop shop for user input and then for output as the configuration file will be modified by the program to hold the results when its done. 
This allows for a self-contained 'file' on the project. Think of it like the doctor keeping a file on you in the office.

## Issues closed:

- Closes #121 
- Closes #111 
- Closes #110

### CODE REVIEW

File: modules/DMZ/project_config/README.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
+# Project Configuration
+_Handles all user input and output_
+
+## Overview
+While there are a many modules & files in this project, as a _user_, I expect there to be a single point of entry where I can go and take care of everything.
+# Purpose
```
Comment (ASAAR) [inline modules/DMZ/project_config/README.md]: The fact that some of these headings have an empty line before them and some don't, is somewhat vexing to me.

File: modules/DMZ/project_config/configuration.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        self.root = self.tree.getroot()
+        if not self.root.tag == TAG_ROOT:
+            raise ConfigurationError('Required XML root tag [{}] not found in\
+             {}'.format(TAG_ROOT, self.config_file_name))
+
+        self.project_name = self.root.attrib.get('name', '')
+        self.user_name = self.root.attrib.get('user', '')
+
+        for child in self.root:
+            #print(child)
```
Comment (ASAAR) [inline modules/DMZ/project_config/configuration.py]: Do you need to keep this?

File: modules/DMZ/project_config/tests/test_configuration.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    Creates dummmy test & train file names and adds to attributes under key:
+    Models and the value is a dictionary of dictionaries with
+    key,value = model name, a dictionary of params where the nest dictionary is
+    key,value = param attribute, param value
+    Each param is guarenteed to have a name & numeric attribute.
+    If number=false, then there will be a value attribute with a list of values.
+    """
+    models = {}
+    for i in range(1, 2):
+        model = {}
```
Comment (RyanMcBerg) [inline modules/DMZ/project_config/tests/test_configuration.py]: doesn't this range only look at 1... range goes from [1, 2)?

File: modules/DMZ/project_config/tests/test_configuration.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    return root
+
+def create_attrib_files(attributes):
+    """
+    To be used after #create_xml_project.
+    Creates dummmy test & train file names and adds to attributes under key:
+    Files and the value is a dictionary with the key,value = path,type
+    """
+    files = {}
+    for i in range(1, 5):
```
Comment (RyanMcBerg) [inline modules/DMZ/project_config/tests/test_configuration.py]: What does this range represent?

File: modules/DMZ/project_config/README.md (hunk 1 above)
Comment (asclines) [inline modules/DMZ/project_config/README.md]: I agree..

File: modules/DMZ/project_config/configuration.py (hunk 2 above)
Comment (asclines) [inline modules/DMZ/project_config/configuration.py]: Nope.

File: modules/DMZ/project_config/tests/test_configuration.py (hunk 4 above)
Comment (asclines) [inline modules/DMZ/project_config/tests/test_configuration.py]: An operator error.

File: modules/DMZ/project_config/tests/test_configuration.py [hunk 5]
```diff
@@ -159,7 +160,7 @@ class TestConfigurationValidXML(TestCase):
     """ Happy Path Testing for XML file """
```
Comment (ASAAR) [inline modules/DMZ/project_config/tests/test_configuration.py]: .......

Comment (asclines) [inline modules/DMZ/project_config/tests/test_configuration.py]: [What?](https://en.wikipedia.org/wiki/Happy_path)

**Pull request merged**